app/javascript/components/student/Editor.tsx

```diff
@@ -17,39 +18,104 @@ export enum TestRunStatus {
   OPS_ERROR = 'ops_error',
 }
 
+type State = {
+  submission?: Submission
+  isSubmitting: boolean
+  code: string
+}
+
+export type Action =
+  | { type: 'submitting'; payload: { code: string } }
+  | { type: 'submitted'; payload: { submission: Submission } }
+  | { type: 'cancelled' }
+
+function reducer(state: State, action: Action): State {
+  switch (action.type) {
+    case 'submitting':
+      return {
+        submission: undefined,
+        code: action.payload.code,
+        isSubmitting: true,
+      }
+    case 'submitted':
+      return {
+        ...state,
+        submission: action.payload.submission,
+        isSubmitting: false,
+      }
+    case 'cancelled':
+      return {
+        ...state,
+        isSubmitting: false,
+      }
+    default:
+      return state
+  }
+}
+
 export function Editor({
   endpoint,
   timeout = 60000,
 }: {
   endpoint: string
   timeout?: number
 }) {
-  const [submission, setSubmission] = useState<Submission>()
+  const [{ submission, isSubmitting, code }, dispatch] = useReducer(reducer, {
+    submission: undefined,
+    isSubmitting: false,
+    code: '',
+  })
   const controllerRef = useRef(new AbortController())
+  const abortFetch = useCallback(() => {
+    controllerRef.current.abort()
+    controllerRef.current = new AbortController()
+  }, [controllerRef.current])
 
   useEffect(() => {
     return () => {
-      controllerRef.current.abort()
+      abortFetch()
     }
   }, [])
 
-  function submit(code: string) {
-    controllerRef.current.abort()
-    controllerRef.current = new AbortController()
-    setSubmission(undefined)
+  useEffect(() => {
+    if (isSubmitting) {
+      return
+    }
+
+    abortFetch()
+  }, [isSubmitting])
+
+  useEffect(() => {
+    if (code === '') {
+      return
+    }
+
+    if (isSubmitting) {
+      abortFetch()
+    }
 
     fetchJSON(endpoint, {
       method: 'POST',
       signal: controllerRef.current.signal,
       body: JSON.stringify({ files: { file: code } }),
-    }).then((json: any) => {
-      setSubmission(typecheck<Submission>(json, 'submission'))
     })
-  }
+      .then((json: any) => {
+        dispatch({
+          type: 'submitted',
+          payload: { submission: typecheck<Submission>(json, 'submission') },
+        })
+      })
+      .catch((responseOrError) => {
+        if (responseOrError.name === 'AbortError') {
+          console.log('Fetch cancelled')
+        }
```

Comment:
```suggestion
        }
        
        throw responseOrError
```

Rethrow, or this error is lost forever.